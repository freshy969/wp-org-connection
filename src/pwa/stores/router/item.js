import { when } from 'mobx';
import { types, getParent, getRoot } from 'mobx-state-tree';

const BaseItem = types
  .model('BaseItem')
  .props({
    mstId: types.identifier(types.string),
    type: types.string,
    id: types.union(types.string, types.number),
    visited: false,
  })
  .views(self => ({
    get connection() {
      return getRoot(self);
    },
    get entity() {
      return self.connection.entity(self.type, self.id);
    },
    get ready() {
      return self.entity.ready;
    },
    get isSingle() {
      return !self.page;
    },
    get isList() {
      return !!self.page;
    },
    get parentColumn() {
      return getParent(self, 2);
    },
    get parentContext() {
      return getParent(self, 4);
    },
    get nextItem() {
      const items = getParent(self);
      const index = items.indexOf(self);
      return index === items.length - 1
        ? self.column.nextColumn && self.column.nextColumn.items[0]
        : items[index + 1];
    },
    isExtracted() {
      return false;
    },
  }));

export const List = BaseItem.named('List')
  .props({
    page: types.number,
    extract: types.maybe(types.enumeration('Extract', ['horizontal', 'vertical'])),
  })
  .views(self => ({
    get list() {
      return self.connection.list(self.type, self.id);
    },
    isExtracted(direction) {
      direction = direction ? [direction] : ['horizontal', 'vertical'];
      return direction.includes(self.extract);
    },
  }))
  .actions(self => {
    let stopReplace = null;
    return {
      beforeDestroy: () => { stopReplace(); },
      afterCreate: () => {
        if (['horizontal', 'vertical'].includes(self.extract)) {
          const { type, id, page, extract } = self;
          stopReplace = when(
            () => self.connection.list(type, id).page(page).ready === true,
            () => {
              self.parentContext.replaceExtractedList({ type, id, page, extract });
            },
          );
        }
      },
    };
  });

export const Single = BaseItem.named('List').props({
  fromList: types.optional(types.frozen, { type: 'latest', id: 'post', page: 1 }),
});

const Item = types.union(({ page }) => (page ? List : Single), List, Single);

export default Item;
