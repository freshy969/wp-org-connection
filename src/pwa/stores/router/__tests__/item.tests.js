import { Item } from '../item';

describe('Router > Context > Column > Item', () => {
  test('Item props are populated appropriately', () => {
    const post = Item.create({
      _id: 'c0dab071-a251-4dac-9801-a2792a3bb362',
      route: 'single',
      singleType: 'post',
      singleId: 60,
      fromList: {
        _id: 'e5571d38-b4c9-4c9e-bb7a-404954cad292',
        listType: 'latest',
      },
    });

    const page = Item.create({
      _id: 'b783bfa6-9ff7-43dd-b7d3-8cf6d215ef7f',
      route: 'single',
      singleType: 'page',
      singleId: 30,
      fromList: {
        _id: '7ee66289-df96-4b2b-a08b-5527cacfc266',
        listType: 'latest',
      },
    });

    const latest = Item.create({
      _id: 'c8bf4c53-370d-4e75-b023-917ec3a4a3b5',
      listType: 'latest',
    });

    const category = Item.create({
      _id: '2fbd1661-41fe-40d2-8df5-009f293ada41',
      listType: 'category',
      listId: '12',
    });

    expect(post).toMatchSnapshot();
    expect(page).toMatchSnapshot();
    expect(latest).toMatchSnapshot();
    expect(category).toMatchSnapshot();
  });
});
