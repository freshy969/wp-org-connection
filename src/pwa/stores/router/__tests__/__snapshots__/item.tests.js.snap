// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router > Context > Column > Item Item props are populated appropriately 1`] = `
Object {
  "_id": "c0dab071-a251-4dac-9801-a2792a3bb362",
  "fromList": Object {
    "_id": "e5571d38-b4c9-4c9e-bb7a-404954cad292",
    "listId": "post",
    "listType": "latest",
    "page": 1,
    "route": "list",
  },
  "route": "single",
  "singleId": 1234,
  "singleType": "post",
}
`;

exports[`Router > Context > Column > Item Item props are populated appropriately 2`] = `
Object {
  "_id": "b783bfa6-9ff7-43dd-b7d3-8cf6d215ef7f",
  "fromList": Object {
    "_id": "7ee66289-df96-4b2b-a08b-5527cacfc266",
    "listId": "post",
    "listType": "latest",
    "page": 1,
    "route": "list",
  },
  "route": "single",
  "singleId": 1234,
  "singleType": "page",
}
`;

exports[`Router > Context > Column > Item Item props are populated appropriately 3`] = `
Object {
  "_id": "c8bf4c53-370d-4e75-b023-917ec3a4a3b5",
  "listId": "post",
  "listType": "latest",
  "page": 1,
  "route": "list",
}
`;

exports[`Router > Context > Column > Item Item props are populated appropriately 4`] = `
Object {
  "_id": "2fbd1661-41fe-40d2-8df5-009f293ada41",
  "listId": "12",
  "listType": "category",
  "page": 1,
  "route": "list",
}
`;
